modules/core/src/smithy4s/Document.scala

```diff
@@ -35,6 +35,17 @@ sealed trait Document extends Product with Serializable {
 
   override def toString(): String = this.show
 
+  def name: String = this match {
+    case DNumber(_)  => "Number"
+    case DString(_)  => "String"
+    case DBoolean(_) => "Boolean"
+    case DNull       => "Null"
+    case DArray(value) =>
+      s"Array[${value.headOption.map(_.name).getOrElse("Null")}]"
```

Comment:
there is absolutely no guarantee that an json array may be homogenous in shape. I'd rather it was just a static `"Array"`, and for object to be `"Object"`.